### PR TITLE
Improved `for lifetime` handling

### PR DESCRIPTION
**Pull request by jdupak**
Tested all commits check-rust + format + commit message.

@P-E-P

### CODE REVIEW

File: gcc/rust/parse/rust-parse-impl.h [hunk 1]
```diff
@@ -4029,6 +4039,12 @@ Parser<ManagedTokenSource>::parse_trait_bound ()
 
   location_t locus = lexer.peek_token ()->get_locus ();
 
+  /* parse for lifetimes, if it exists (although empty for lifetimes is ok to
+   * handle this) */
```
[the remark this answers is not in the document]
Comment (jdupak) [reply gcc/rust/parse/rust-parse-impl.h:4043]: Changed to `/* parse optional `for lifetimes`. */`

**Comment by jdupak** [comment]
@CohenArthur fixed and added a test

**Comment by CohenArthur** [comment]
@jdupak this looks good to me, can you investigate why the tests are failing? since this change concerns the parser mainly, you can add `// { dg-additional-options "-fsyntax-only" }` to the testcase so it stops after parsing

**Comment by jdupak** [comment]
@CohenArthur fixed

**Review by CohenArthur — APPROVED** [review]
LGTM :)

**Pull request merged**